### PR TITLE
`Adaptive learning`: Extend learning path feature to recommend sequence of lecture units and exercises

### PR DESCRIPTION
**Pull request by MaximilianAnzinger**
⚠️ Stacked on #7091  please don't merge accidentally
======

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

We want to provide a sequential order of lecture units and exercises for students participating via the learning paths feature.

### Description

This PR adds the second half for generating this sequential order by recommending an order of learning objects within the competencies.
The LearningPathService is now split into the `LearningPathService` that should be used by other components and the sub-services `LearningPathRecommendationService` and `LearningPathNgxSercive` that handle the corresponding sub-tasks.
In addition to recommending an order of competencies, the recommendation now includes a suggested order of the learning objects (lecture units and exercises). The recommender will select exercises by difficulty level according to the prior performance of the student and also recommend repetitions of already completed exercises if the achieved score is comparably low.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Course with at least one competency and learning paths enabled
- exercises and lecture units for each type linked to competency

1. Log in to Artemis as Student
2. Navigate to the Learning Path participation view in the course
3. Make sure the visualized graph on the left side of your screen is displayed as expected (no mastered competencies displayed, the competencies are ordered sequentially/a chain from top to bottom excluding the nodes for the learning objects, and the recommended lecture units/exercises are a linear chain between the competency start and end nodes)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [CompetencyProgressService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5351/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/CompetencyProgressService.html) | 91% | ✅  |
| [CourseService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5351/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/CourseService.html) | 89% | ✅ |
| [ExerciseService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5351/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/ExerciseService.html) | 96% | ✅  |
| [LearningObjectService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5351/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/LearningObjectService.html) | 89% | ✅  |
| [ParticipantScoreService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5351/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/ParticipantScoreService.html) | 96% | ✅  |
| [LearningPathNgxService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5351/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.learningpath/LearningPathNgxService.html) | 97% | ✅  |
| [LearningPathRecommendationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5351/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.learningpath/LearningPathRecommendationService.html) | 98% | ✅  |
| [LearningPathService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5351/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.learningpath/LearningPathService.html) | 98% | ✅  |
| [CompetencyResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5351/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/CompetencyResource.html) | 97% | ✅  |
| [CourseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5351/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/CourseResource.html) | 94% | ✅  |
| [LearningPathResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5351/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/LearningPathResource.html) | 91% | ✅  |

### Screenshots

As this PR doesn't change the UI there are no screenshots of shiny new components, just a comparison between the old graph received from the server (in participation view) and the new one:

Previously:
Screenshot 2023-08-31 at 11 47 07

Now (excludes completed exercises and displays linear path):
Screenshot 2023-08-31 at 11 52 00

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:375]:
```suggestion
        List<LearningObject> recommendedOrder = new ArrayList<>(pendingLectureUnits);
```
Did you include the `Array` on purpose?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:386]: Forgotten log?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:387]:
```suggestion
        Map<DifficultyLevel, Set<Exercise>> difficultyLevelMap = generateDifficultyLevelMap(competency.getExercises());
```
Same as for the list above

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:394]: Why can this happen now but was not caught in the if before? Because the exercises have the "wrong" difficulty and you predict that you cannot reach the mastery with them?
Maybe explain this a bit in a comment

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:409]: I've seen a variation of this formula and its derivatives in a few PRs now. Is it possible to move all related computations to one place, so it can be easily changed afterwards?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:435]: Would it make sense to use the average score of the prior competencies instead of an arbitrary threshold of 50%?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:445]: Should the last parameter always be `easyExercises`?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:461]: Why did you chose the order medium > easy > hard?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:501]: Why do you sort before computing the average?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:526]: Maybe make team scale happy and use lower case names?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:528]: What is `LO` supposed to mean? Can you use longer names?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:575]:
```suggestion
        final var numberOfExercises = new int[DifficultyLevel.values().length];
```
Shouldn't matter, but it's clearer where the 3 comes from and one less place to potentially change in the future

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/LearningObjectService.java
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/LearningObjectService.java:16]: Javadoc would be nice, and maybe a few words what a learning object is because i've never heard that term in the artemis context.

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:80]: why do we return the state here? shouldn't be the state of the recommendation always be finished or something similar if we return the recommended order?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:97]: any particular reason you do not use map as static type?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:150]: can we not just return a map?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:131]: why does it need to be a hashmap?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:477]: missing exercises for what? I assume the missing number of exercises that need to be selected with a particular difficulty?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:59]: maybe state once that LUT means lookup table because at least I didn't know that and it's annoying if you have to google abbreviations to understand the code.

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (tobias-lippert) [reply src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:526]: Is my assumption correct that LU means lecture unit?

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (tobias-lippert) [reply src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:528]: I guess it means learning objects but I aggree that it's really not descriptive.

File: src/test/java/de/tum/in/www1/artemis/StudentScoreUtilService.java
Comment (tobias-lippert) [inline src/test/java/de/tum/in/www1/artemis/StudentScoreUtilService.java:16]: can this be package private?

File: src/test/java/de/tum/in/www1/artemis/StudentScoreUtilService.java
Comment (tobias-lippert) [inline src/test/java/de/tum/in/www1/artemis/StudentScoreUtilService.java:19]: any reason the fields are not private?

**Review by tobias-lippert — COMMENTED** [review]
Code looks good in general, as always I have a few questions and would prefer variable names that are longer than 1-3 characters.

**Review by lennart-keller — APPROVED** [review]
Tested in testing session, worked well

File: src/test/java/de/tum/in/www1/artemis/StudentScoreUtilService.java
Comment (MaximilianAnzinger) [reply src/test/java/de/tum/in/www1/artemis/StudentScoreUtilService.java:15]: All our util services are intended to be useable for all our integration testing needs. Therefore, all of them are public.

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java
Comment (MaximilianAnzinger) [reply src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java:74]: I want to reuse the prior mapping rather than reconstructing it again. I could use another record to return the mapping and the recommended order.

Let me know if you want me to change this or if that's fine with you.

**Review by JohannesStoehr — APPROVED** [review]
Approve after Tobias‘ and my change requests

**Review by tobias-lippert — APPROVED** [review]
Code LGTM now.

**Review by nityanandaz — DISMISSED** [review]
Manual test on TS2 (2023-09-07)

The graph is in sequential order

**Review by Kroko-fant — APPROVED** [review]
Tested in a testing session on ts2 the path appears as it should appear

File: src/test/java/de/tum/in/www1/artemis/competency/CompetencyUtilService.java
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/competency/CompetencyUtilService.java:136]: codacy warning: missing javadoc tag.

**Review by JohannesStoehr — APPROVED** [review]
Reapprove Code after merges

File: src/test/java/de/tum/in/www1/artemis/service/LearningPathServiceTest.java
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/service/LearningPathServiceTest.java:583]:
```suggestion
            assertThat(actual.nodes()).hasSize(6);
```

**Review by Strohgelaender — APPROVED** [review]
code lgtm

**Review by JohannesStoehr — APPROVED** [review]

**Review by florian-glombik — APPROVED** [review]
Tested on ts1, the described use case worked fine (split path is now displayed sequentially)

A view points out of scope for this PR:
* the continue button was always disabled (not quite sure if the up down arrow solution is in another PR)
* I did not understand how to delete and modify competency relations
* Deleting a competency that is linked via relation did not work (`400` as response)

**Comment by MaximilianAnzinger** [comment]
@florian-glombik thank you for your review 👍 

> * the continue button was always disabled (not quite sure if the up down arrow solution is in another PR)

the continue button & sidebar is reworked and fixed in #7235 

> * I did not understand how to delete and modify competency relations

this is unrelated to this PR, but you can click on the respective edge and delete it that way.

> * Deleting a competency that is linked via relation did not work (`400` as response)

do you know if this is also the case on develop? I'll take a look at it but this is an unrelated issue.

**Pull request merged**
